csharp/src/Ice/ConnectionI.cs

```diff
@@ -1902,29 +1902,44 @@ private bool validate(int operation)
         return true;
     }
 
+
+    /// <summary>
+    /// Sends the next queued messages. This is called by message() when the message from _writeStream was fully sent so
+    /// before sending the next message the message which was being sent (_sendStreams.First) is first removed from the
+    /// queue and if it has a sent callback to call, the sent callback is queued.
+    /// </summary>
+    /// <param name="callbacks">The sent callbacks to call for the messages that were sent.</param>
+    /// <returns></returns>
     private int sendNextMessage(out Queue<OutgoingMessage> callbacks)
     {
         callbacks = null;
 
         if (_sendStreams.Count == 0)
         {
+            // This can occur if no message was being written and the socket write operation was registered with the
+            // thread pool (a transceiver read method can request writing data).
             return SocketOperation.None;
         }
         else if (_state == StateClosingPending && _writeStream.pos() == 0)
         {
-            // Message wasn't sent, empty the _writeStream, we're not going to send more data.
+            // Message wasn't sent, empty the _writeStream, we're not going to send more data because the connection
+            // is being closed.
             OutgoingMessage message = _sendStreams.First.Value;
             _writeStream.swap(message.stream);
             return SocketOperation.None;
         }
 
+        // Assert that the message was fully written.
         Debug.Assert(!_writeStream.isEmpty() && _writeStream.pos() == _writeStream.size());
+
         try
         {
             while (true)
             {
                 //
-                // Notify the message that it was sent.
+                // The message that was being sent is sent so we can swap back the write stream buffer to the
+                // outgoing message (required for retry) and queue its sent callback if there's a sent callback to
+                // call.
```

Comment:
After line 1945, is _writeStream empty? If not, what is in _writeStream?

Reply:
Yes, `_writeStream` is empty. I've added documentation to the `_readStream` and `_writeStream` data members.